packages/library/src/commonMain/kotlin/io/realm/internal/SuspendableNotifier.kt

```diff
@@ -176,73 +177,88 @@ internal class SuspendableNotifier(private val owner: Realm, private val dispatc
+        }
+        return registerChangedListener(
+            liveComponentPointer = liveObject.`$realm$ObjectPointer`!!,
+            notifyComponentUpdate = { frozenRealm ->
+                if (!liveObject.isValid()) {
+                    callback.onChange(null)
+                } else {
+                    callback.onChange(liveObject.freeze(frozenRealm))
+                }
+            },
+            getToken = { objPtr, interopCallback ->
+                RealmInterop.realm_object_add_notification_callback(objPtr, interopCallback)
+            },
+            callback = callback
+        )
     }
 
-    // FIXME Need to expose change details to the user
-    //  https://github.com/realm/realm-kotlin/issues/115
     /**
-     * Listen to changes to a RealmObject through a change listener. The callback will happen
+     * Listen to changes to a RealmList through a change listener. The callback will happen
      * on the configured [SuspendableNotifier.dispatcher] thread.
      *
      * FIXME Callers of this method must make sure it is called on the correct [SuspendableNotifier.dispatcher].
      */
-    internal fun <T : RealmObject> registerObjectChangedListener(obj: T, callback: Callback<T?>): Cancellable {
-        val liveObject: RealmObjectInternal? = (obj as RealmObjectInternal).thaw(realm.realmReference.owner) as RealmObjectInternal?
-        if (liveObject == null || !liveObject.isValid()) {
-            return NO_OP_NOTIFICATION_TOKEN
-        }
-        val token = RealmInterop.realm_object_add_notification_callback(
-            liveObject.`$realm$ObjectPointer`!!,
-            object : io.realm.interop.Callback {
-                override fun onChange(objectChanges: NativePointer) {
-                    // FIXME How to make sure the Realm isn't closed when handling this?
-
-                    // FIXME The Realm should have been frozen in `realmChanged`, but this isn't supported yet.
-                    //  Instead we create the frozen version ourselves (which is correct, but pretty inefficient)
-                    val frozenRealm = RealmReference(owner, RealmInterop.realm_freeze(realm.realmReference.dbPointer))
-                    notifyRealmChanged(frozenRealm)
-
-                    if (!liveObject.isValid()) {
-                        callback.onChange(null)
-                    } else {
-                        callback.onChange(liveObject.freeze(frozenRealm))
-                    }
+    internal fun <T> registerListChangedListener(
+        list: RealmList<T>,
+        callback: Callback<RealmList<T>>
+    ): Cancellable {
+        val liveList = list.thaw(realm.realmReference)
+        return registerChangedListener(
+            liveComponentPointer = liveList.listPtr,
+            notifyComponentUpdate = { frozenRealm ->
+                // Skip notifications when list is not valid
+                if (!liveList.isValid()) {
+                    return@registerChangedListener
```

Comment:
The behavior here is different than for Objects. We should make them consistent. I.e. also return `null` here. That will probably result in the flow completing further up, but we will also have a normal callback based API on top of this, where we probably want to expose the `null` directly.